### PR TITLE
Extract PushMetricsMarker as separate gem

### PR DESCRIPTION
**Pull request by aelkiss**
Extracts PushMetricsMarker and its specs from
https://github.com/hathitrust/holdings-backend/blob/main/lib/utils/push_metrics_marker.rb

I decided to make a separate gem rather than adding to milemarker:

- it doesn't make sense to require prometheus-client as a dependency for milemarker (it isn't required to use the basic functionality)
- it doesn't make sense to require things that use this to separately declare prometheus-client as a dependency in order for it to work

### CODE REVIEW

**Comment by aelkiss** [comment]
Still to do:

* README/documentation
* github actions setup (will adapt https://github.com/mlibrary/tkseal/tree/main/.github/workflows)

**Comment by aelkiss** [comment]
Added tests workflow. 

As far as releasing, https://github.com/mlibrary/tkseal/blob/main/.github/workflows/release.yml relies on "conventional commits", which I'm not sure I want to adopt yet. To start, I think we can just point to the gem sources in github, and figure out publishing later.

**Comment by aelkiss** [comment]
@billdueber Mainly just want to make sure this looks like a reasonable setup, documentation, & that it makes sense to you to keep this separate from milemarker.

@mwarin For you, I mainly want to check that this makes sense as an extraction of https://github.com/hathitrust/holdings-backend/blob/main/lib/utils/push_metrics_marker.rb

File: lib/push_metrics.rb [hunk 1]
```diff
@@ -0,0 +1,77 @@
+# frozen_string_literal: true
+
+require "milemarker"
+require "delegate"
+
+# Adds prometheus push exporter to Milemarker that tracks:
+#   - number of items processed
+#   - time running so far
+#   - success time
+class PushMetrics < SimpleDelegator
```
Comment (billdueber) [inline lib/push_metrics.rb:10]: ~I see `SimpleDelegator`, but don't see a `__setobj__` call anywhere -- did you decide to not delegate, or just forget to call it?~

TIL that `SimpleDelegator.new(obj)` sets the target object, here via `super`.

File: lib/push_metrics.rb [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+  def update_metrics
+    duration_metric.set(@marker.total_seconds_so_far)
+    records_processed_metric.set(@marker.count)
+
+    pushgateway.add(registry)
+  end
+
+  def duration_metric
+    @duration_metric ||= register_metric(:job_duration_seconds, docstring: "Time spend running job in seconds")
```
Comment (billdueber) [inline lib/push_metrics.rb]: "spent"

File: lib/push_metrics.rb [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    registry: Prometheus::Client.registry,
+    job_name: File.basename($PROGRAM_NAME),
+    pushgateway_endpoint: ENV["PUSHGATEWAY"] || "http://localhost:9091",
+    success_interval: ENV["JOB_SUCCESS_INTERVAL"],
+    pushgateway: Prometheus::Client::Push.new(job: job_name, gateway: pushgateway_endpoint))
+
+    @marker = marker
+    @pushgateway = pushgateway
+    @registry = registry
+    @metrics = metrics
```
Comment (billdueber) [inline lib/push_metrics.rb]: I can't figure out where `metrics` is coming from.  Maybe leftover? It looks like metrics are all defined within the `registry` via side-effect calls to `register_metric`?

**Review by billdueber — APPROVED** [review]
At first blush, this looks great. Two minor comments.

File: lib/push_metrics.rb (hunk 3 above)
Comment (aelkiss) [reply lib/push_metrics.rb]: The effect at initialization time (with `attr_reader :metrics`) is to... set `@metrics` to `nil`. Good catch. Appears to be leftover from https://github.com/hathitrust/holdings-backend/commit/91b3837454bf8c64910b4739bb9a7be689275553#diff-5ec2879cbfa80ffff097f34259d77c0aa89d950cbfcf98a66d0392953f3e0e5a which moved the metrics registration into this class - I think that was a change for prometheus client 4.0.0, which automatically pushed any registered metrics - so we needed to change to lazily register them instead

**Pull request merged**